uniffi/src/templates/ObjectTemplate.rs

```diff
@@ -40,25 +40,51 @@ lazy_static::lazy_static! {
 {% endfor %}
 
 {%- for meth in obj.methods() %}
-    #[no_mangle]
+
+{% match meth.throws() %}
+    {% when Some with (e) %} 
+#[no_mangle]
     pub extern "C" fn {{ meth.ffi_func().name() }}(
         {%- for arg in meth.ffi_func().arguments() %}
         {{ arg.name() }}: {{ arg.type_()|type_c }},
         {%- endfor %}
+        err: &mut ffi_support::ExternError,
+    ) -> {% match meth.ffi_func().return_type() %}{% when Some with (return_type) %}{{ return_type|ret_type_c }}{% else %}(){% endmatch %} {
+        log::debug!("{{ meth.ffi_func().name() }}");
+        // If the method does not have the same signature as declared in the IDL, then
+        // this attempt to call it will fail with a (somewhat) helpful compiler error.
+        UNIFFI_HANDLE_MAP_{{ obj.name()|upper }}.call_with_result_mut(err, {{ meth.first_argument().name() }}, |obj| -> Result<{% match meth.ffi_func().return_type() %}{% when Some with (return_type) %}{{ return_type|ret_type_c }}{% else %}(){% endmatch %}, {{ e }}> {
+            let _retval = {{ obj.name() }}::{{ meth.name() }}(
+                obj,
+                {%- for arg in meth.arguments() %}
+                {{ arg.name()|lift_rs(arg.type_()) }},
+                {%- endfor %}
+            )?;
+            Ok({% match meth.return_type() %}{% when Some with (return_type) %}{{ "_retval"|lower_rs(return_type) }}{% else %}_retval{% endmatch %})
+        })
+    }
+
+    {% else %}
+#[no_mangle]
+    pub extern "C" fn {{ meth.ffi_func().name() }}(
+        {%- for arg in meth.ffi_func().arguments() %}
+        {{ arg.name() }}: {{ arg.type_()|type_c }},
+        {%- endfor %}
+        err: &mut ffi_support::ExternError,
```

Comment:
Since you're always taking an `err` param here even if the underlying method can't throw, you could consider pushing the `{% match meth.throws() %}` down into the  body of the method, and using to ensure that we always return a `Result`. Something like:

```
        UNIFFI_HANDLE_MAP_foo.call_with_result_mut(err, [...], |obj| -> Result<[...], [... can we use "_" here to infer the error type? ...]> {
            let _retval = {{ obj.name() }}::{{ meth.name() }}(
                obj,
                {%- for arg in meth.arguments() %}
                {{ arg.name()|lift_rs(arg.type_()) }},
                {%- endfor %}
            );

            {% match meth.throws() %}
            {% when Some with (e) %} 
            let _retval = _retval?;
            {% endmatch %}

            Ok([... cast "_retval" into ffi type...])
        })
    }
```

Not sure if that would make it easier or harder to follow...you might be better off just branching on "does it throw?" and "does it have a return type" and doing four different templates for four different combos.